### PR TITLE
fix(ct): Don't bloat metainfo.json file when # or ? is present in id

### PR DESCRIPTION
**Pull request by mikstime**
## Proposal

Remove the `?` or `#` part from dependencies in the `metainfo.json` file.

## Description

`@playwright/experimental-ct-core` generates a `metainfo.json` file used for caching and other optimizations performed by Playwright under the hood.

The current implementation relies on a Vite plugin to generate this file. However, Vite prohibits the usage of `#` and `?` in file names, meaning there is no reason to store them in the `metainfo.json` file.

> Both # and ? are reserved for compatibility with browser URLs.
>
> -- <cite><a href="https://github.com/vitejs/vite/issues/11784#issuecomment-1798854088">patak-dev</a></cite>

We experienced two major issues with the current behavior:

- Our `metainfo.json` file exceeded 512MB, which is the hard limit for the `JSON.parse()` method, resulting in unexpected crashes

- This data was serialized and deserialized each time tests ran, resulting in a 4x slowdown of tests in our project

80% performance improvement with proposed fix

## Example

Metainfo.json file is reduced from 1.4M to 4.0K in provided example

```
1.4M    current/playwright/.cache/metainfo.json
4.0K    patched/playwright/.cache/metainfo.json
```
https://github.com/mikstime/playwright

### CODE REVIEW

**Comment by dgozman** [comment]
@mikstime Thank you for the PR! Is there any way you can share a repro, or even better - turn that into a test? That would help us to review and prevent future regressions.

**Comment by mikstime** [comment]
> @mikstime Thank you for the PR! Is there any way you can share a repro, or even better - turn that into a test? That would help us to review and prevent future regressions.

Here is an example https://github.com/mikstime/playwright

**Comment by dgozman** [comment]
> > @mikstime Thank you for the PR! Is there any way you can share a repro, or even better - turn that into a test? That would help us to review and prevent future regressions.
> 
> Here is an example https://github.com/mikstime/playwright

Wonderful! Could you please add a test to `tests/playwright-test/playwright.ct-build.spec.ts` that follows this example? You can run it with `npm run ttest playwright.ct-build.spec:<line>`.

**Comment by mikstime** [comment]
@microsoft-github-policy-service agree

**Comment by mikstime** [comment]
> > > @mikstime Thank you for the PR! Is there any way you can share a repro, or even better - turn that into a test? That would help us to review and prevent future regressions.
> > 
> > 
> > Here is an example https://github.com/mikstime/playwright
> 
> Wonderful! Could you please add a test to `tests/playwright-test/playwright.ct-build.spec.ts` that follows this example? You can run it with `npm run ttest playwright.ct-build.spec:<line>`.

Sure. Thank you for the instructions.
```
Running 1 test using 1 worker

  ✓  1 [playwright-test] › tests/playwright-test/playwright.ct-build.spec.ts:740:5 › should remove ? or # from deps in metainfo.json (3.1s)

  1 passed (3.8s)
```

File: packages/playwright-ct-core/src/vitePlugin.ts [hunk 1]
```diff
@@ -291,7 +291,7 @@ function vitePlugin(registerSource: string, templateDir: string, buildInfo: Buil
 function collectViteModuleDependencies(context: PluginContext, id: string, deps: Set<string>) {
   if (!path.isAbsolute(id))
     return;
-  const normalizedId = path.normalize(id);
+  const normalizedId = path.normalize(id).split(/\?|#/)[0];
```
Comment (dgozman) [inline packages/playwright-ct-core/src/vitePlugin.ts]: - I think we should strip `?` and `#` before the line 292, just in case there was something like `/..` in the query params or hash.
- It would be great to add a comment explaining/having an example of an import with a query param.

File: tests/playwright-test/playwright.ct-build.spec.ts [hunk 2]
```diff
@@ -736,3 +736,62 @@ test('should resolve components imported from node_modules', async ({ runInlineT
   expect(result.exitCode).toBe(0);
   expect(result.passed).toBe(1);
 });
+
+test('should remove ? or # from deps in metainfo.json', async ({ runInlineTest }, testInfo) => {
```
Comment (dgozman) [inline tests/playwright-test/playwright.ct-build.spec.ts:740]: Thank you for the test, this really helped!

Just to understand the scenario better, do you have a custom loader/plugin/something else that handles query params? What's the actual usecase of yours, not the toy example we have here?

Can we have a test that resembles real life? IIUC, this should manifest in watch not triggering when you change a styles file that was imported as `'./style.css?inline'`. At least, that's the impression I get from reading vite docs. If so, we can add a simple test next to `should watch component when editing util`.

File: packages/playwright-ct-core/src/vitePlugin.ts (hunk 1 above)
Comment (mikstime) [reply packages/playwright-ct-core/src/vitePlugin.ts]: Fixed. I’ve stripped ? and # before line 292 and added a comment with an example of an import that includes query parameters.
For context: even without stripping, path.isAbsolute would still treat such paths as absolute. According to the Node.js [documentation](https://nodejs.org/api/path.html#pathisabsolutepath), paths like /baz/.. are considered absolute, and this also applies when query params or hashes are present, e.g.:

```
path.isAbsolute('C:/foo/..');   // true
path.isAbsolute('C:\\foo\\..'); // true
path.isAbsolute('/baz/..');    // true
path.isAbsolute('/baz/../..'); // true
```

This is also true:
```
> p.isAbsolute('/dir/file#x=../../b.png')
true
> p.isAbsolute('/dir/file?x=../../b.png')
true
```

That said, removing `?` and `#` up front makes the intent clearer and avoids any edge cases involving traversal in query or hash parts.

File: tests/playwright-test/playwright.ct-build.spec.ts (hunk 2 above)
Comment (mikstime) [reply tests/playwright-test/playwright.ct-build.spec.ts:740]: > Just to understand the scenario better, do you have a custom loader/plugin/something else that handles query params? What's the actual usecase of yours, not the toy example we have here?

We use a Vite plugin that transforms source code and appends query parameters to module imports. These query params are used to alter compilation behavior based on the user agent.

> If so, we can add a simple test next to should watch component when editing util.

Sure. I’ve added two tests:
- One that directly covers the query-param import case (e.g. style.css?inline), matching the real-world scenario described above.
- Another that exercises the same behavior through a story file, since playwright compiles these cases differently.

**Review by dgozman — APPROVED** [review]
This looks great, thank you for following through!

**Comment by dgozman** [comment]
@mikstime It seems like this broke our [component tests](https://github.com/microsoft/playwright/actions/runs/21879748147/job/63251225324?pr=39074). Could you please take a look? You can run `npm run ct` for all component tests, or `cd tests/components/ct-svelte && npm install && npx playwright test` to see what's going on with specific tests. Please let me know if you stumble upon any issue.

**Comment by mikstime** [comment]
> @mikstime It seems like this broke our [component tests](https://github.com/microsoft/playwright/actions/runs/21879748147/job/63251225324?pr=39074). Could you please take a look? You can run `npm run ct` for all component tests, or `cd tests/components/ct-svelte && npm install && npx playwright test` to see what's going on with specific tests. Please let me know if you stumble upon any issue.

Sorry about that — looks like I dropped a small but important guard during the rebase:
```
  if (deps.has(normalizedId))
    return;
```

Must be working now. `npm run ct` is passing now _(`18 passed (4.3m)`)_

**Comment by mikstime** [comment]
> ### [Test results](https://mspwblobreport.z1.web.core.windows.net/run-21910485490-2-7f56c1b15b510c98800451f2d20d43fb3560b700/index.html) for "tests 1"
> **4 failed** ❌ [playwright-test] › reporter-html.spec.ts:304 › merged › should not include image diff with non-images `@macos-latest-node20` ❌ [playwright-test] › ui-mode-test-network-tab.spec.ts:371 › should not preserve selection across test runs `@macos-latest-node20` ❌ [playwright-test] › update-aria-snapshot.spec.ts:299 › should update missing snapshots in tsx `@macos-latest-node20` ❌ [playwright-test] › update-aria-snapshot.spec.ts:669 › update-source-method › should overwrite source when specified in the config `@macos-latest-node20`
> 
> 5 flaky
> **38565 passed, 843 skipped**
> 
> Merge [workflow run](https://github.com/microsoft/playwright/actions/runs/21915237781).

Can't reproduce locally on macOS + node 20. Can it be a fluke?

<details>

<summary>Tests passing locally</summary>

```
mikstime@mikstime-osx playwright-1 % node -v
v20.13.1
mikstime@mikstime-osx playwright-1 % npm run ttest update-aria-snapshot.spec.ts:299
mikstime@mikstime-osx playwright-1 % npm run ttest update-aria-snapshot.spec.ts:299

> playwright-internal@1.59.0-next ttest
> node ./tests/playwright-test/stable-test-runner/node_modules/@playwright/test/cli test --config=tests/playwright-test/playwright.config.ts update-aria-snapshot.spec.ts:299


Running 1 test using 1 worker

  ✓  1 [playwright-test] › tests/playwright-test/update-aria-snapshot.spec.ts:299:5 › should update missing snapshots in tsx (10.3s)

  1 passed (11.0s)
mikstime@mikstime-osx playwright-1 % npm run ttest update-aria-snapshot.spec.ts:669

> playwright-internal@1.59.0-next ttest
> node ./tests/playwright-test/stable-test-runner/node_modules/@playwright/test/cli test --config=tests/playwright-test/playwright.config.ts update-aria-snapshot.spec.ts:669


Running 1 test using 1 worker

  ✓  1 [playwright-test] › tests/playwright-test/update-aria-snapshot.spec.ts:669:7 › update-source-method › should overwrite source when specified in the config (8.9s)

  1 passed (9.6s)
mikstime@mikstime-osx playwright-1 % npm run ttest reporter-html.spec.ts:304       

> playwright-internal@1.59.0-next ttest
> node ./tests/playwright-test/stable-test-runner/node_modules/@playwright/test/cli test --config=tests/playwright-test/playwright.config.ts reporter-html.spec.ts:304


Running 2 tests using 2 workers

  ✓  1 [playwright-test] › tests/playwright-test/reporter-html.spec.ts:304:5 › created › should not include image diff with non-images (2.7s)
  ✓  2 [playwright-test] › tests/playwright-test/reporter-html.spec.ts:304:5 › merged › should not include image diff with non-images (3.1s)

  2 passed (5.7s)
mikstime@mikstime-osx playwright-1 % npm run ttest ui-mode-test-network-tab.spec.ts:371

> playwright-internal@1.59.0-next ttest
> node ./tests/playwright-test/stable-test-runner/node_modules/@playwright/test/cli test --config=tests/playwright-test/playwright.config.ts ui-mode-test-network-tab.spec.ts:371


Running 1 test using 1 worker

  ✓  1 [playwright-test] › tests/playwright-test/ui-mode-test-network-tab.spec.ts:371:5 › should not preserve selection across test runs (7.4s)

  1 passed (8.2s)
```

</details>

**Review by dgozman — APPROVED** [review]
All the related tests are passing, merging in. Thank you for the fix!
